Mathlib/GameTheory/Auction/Secondprice.lean

```diff
@@ -0,0 +1,290 @@
+lemma exists_max : ∃ i : a.I, b i = a.maxb b := by
+   obtain ⟨i, _, h2⟩ := Finset.exists_mem_eq_sup' Finset.univ_nonempty b
+   exact ⟨i, symm h2⟩
+
+/-- winner: The participant with the highest bid. -/
+noncomputable def winner : a.I := Classical.choose (exists_max b)
+
+/-- The bid of the winner equals the highest bid. -/
+lemma winner_take_max : b (winner b) = maxb b := Classical.choose_spec (exists_max b)
+
+/-!  ## Finset Nontriviality Lemmas -/
+
+/--  Relates the nontriviality of type `α` with its universal finite set, implying the universal set
+has at least two distinct elements if `α` does. -/
+lemma Finset.univ_nontrivial_iff {α : Type*} [Fintype α] :
+    (Finset.univ : Finset α).Nontrivial ↔ Nontrivial α := by
+  rw [Finset.Nontrivial, Finset.coe_univ, Set.nontrivial_univ_iff]
+
+/-- Asserts that the universal set of a nontrivial type `α` is nontrivial. -/
+lemma Finset.univ_nontrivial {α : Type*} [Fintype α] [h : Nontrivial α] :
+    (Finset.univ : Finset α).Nontrivial :=
+  Finset.univ_nontrivial_iff.mpr h
+
+/-- Proves that the universal set contains distinct elements for any nontrivial type `α`. -/
+lemma Finset.Nontrivial.univ {α} [Fintype α] [Nontrivial α] :
+    Finset.Nontrivial (Finset.univ : Finset α) := by
+  let ⟨a, b, h⟩ := exists_pair_ne α
+  exact ⟨a, Finset.mem_univ _, b, Finset.mem_univ _, h⟩
+
+/-- `B i` is the maximal bid of all participants but `i`. -/
+noncomputable def B (i : a.I) : ℝ := Finset.sup' (Finset.erase Finset.univ i)
+(Finset.Nontrivial.erase_nonempty (Finset.univ_nontrivial)) b
+
+/--The second highest bid: the highest bid excluding the winner’s bid.-/
+noncomputable def secondPrice : ℝ := B b (winner b)
+
+/-- The bid of the winner is always greater than or equal to the bids of all other participants.-/
+lemma b_winner_max (j : a.I) : b j ≤ b (winner b) := by
+  rw [winner_take_max b]
+  exact Finset.le_sup' b (Finset.mem_univ j)
+
+/-- If `i`'s bid is higher than all other bids, then `i` is the winner. -/
+lemma gt_wins (i : a.I) (H : ∀ j , i ≠ j → b j < b i) : i = winner b := by
+  contrapose! H
+  exact ⟨winner b, H, b_winner_max b i⟩
+
+/-- The bid of the winner is always greater than or equal to the second highest bid. -/
+lemma b_winner : secondPrice b ≤ b (winner b) := by
+  rw [winner_take_max b]
+  exact Finset.sup'_mono b (Finset.subset_univ _)
+    (Finset.Nontrivial.erase_nonempty (Finset.univ_nontrivial))
+
+/-- If `i` is not the winner, then the highest bid excluding `i` is equal to the highest bid. -/
+lemma b_loser_max {i : a.I} (H : i ≠ winner b) : B b i = maxb b := by
+  have H1: B b i ≤ maxb b := by
+    apply Finset.sup'_le
+    intro i _
+    apply Finset.le_sup'
+    simp only [Finset.mem_univ]
+  have H2: maxb b ≤ B b i := by
+    rw [← winner_take_max b]
+    apply Finset.le_sup'
+    simp only [Finset.mem_univ, Finset.mem_erase, and_true]
+    exact (H.symm)
+  exact Real.partialOrder.proof_4 (B b i) (Finset.sup' Finset.univ maxb.proof_1 b) H1 H2
```

Comment:
Here's a golf (using `Finset.sup'_mono` to simplify the proof of `H1` and using `le_antisymm` instead of `Real.partialOrder.proof_4`).
```suggestion
  apply le_antisymm
  · apply Finset.sup'_mono
    exact Finset.erase_subset i Finset.univ
  · rw [← winner_take_max b]
    apply Finset.le_sup'
    exact Finset.mem_erase_of_ne_of_mem H.symm (Finset.mem_univ (winner b))
```